unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/PluginSystem/PluginSystem.cs

```diff
@@ -0,0 +1,192 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Assertions;
+
+namespace DCL
+{
+    public delegate IPlugin PluginBuilder();
+
+    /// <summary>
+    /// This class implements a plugin system pattern.
+    /// 
+    /// - Many plugins can share the same feature flag
+    /// - Plugins are registered by using a PluginBuilder delegate that must create and return the plugin instance.
+    /// - Any active plugin is an instantiated plugin. A disabled plugin is a never created or disposed plugin.
+    /// - No Initialize() pattern. Plugins abide to RAII idiom.
+    /// - Feature flags can be set automatically by using SetFeatureFlagsData.
+    /// </summary>
+    public class PluginSystem : IDisposable
+    {
+        private static ILogger logger = new Logger(Debug.unityLogger);
+        private PluginGroup allPlugins = new PluginGroup();
+        private Dictionary<string, PluginGroup> pluginGroupByFlag = new Dictionary<string, PluginGroup>();
+        private BaseVariable<FeatureFlag> featureFlagsDataSource;
+
+        /// <summary>
+        /// Returns true if the plugin defined by the PluginBuilder delegate is currently enabled.
+        ///
+        /// An enabled plugin is an instantiated and currently active plugin.
+        /// </summary>
+        /// <param name="pluginBuilder">The PluginBuilder delegate instance used to instance this plugin.</param>
+        /// <returns>True if the plugin defined by the PluginBuilder delegate is currently enabled</returns>
+        public bool IsEnabled(PluginBuilder pluginBuilder)
+        {
+            if (!allPlugins.plugins.ContainsKey(pluginBuilder))
+                return false;
+
+            return allPlugins.plugins[pluginBuilder].isEnabled;
+        }
+
+        /// <summary>
+        /// Registers a plugin builder and binds it to a feature flag.
+        /// </summary>
+        /// <param name="pluginBuilder">The builder used to construct the plugin.</param>
+        /// <param name="featureFlag">The flag to be bounded. When this flag is true, the plugin will be created.</param>
+        public void RegisterWithFlag(PluginBuilder pluginBuilder, string featureFlag)
+        {
+            Register(pluginBuilder, false);
+            BindFlag(pluginBuilder, featureFlag);
+        }
+
+        /// <summary>
+        /// Registers a new plugin to be built by the PluginBuilder delegate.
+        /// </summary>
+        /// <param name="pluginBuilder">The pluginBuilder instance. This instance will create the plugin when enabled.</param>
+        /// <param name="enable">If true, the plugin will be constructed as soon as this method is called.</param>
+        public void Register(PluginBuilder pluginBuilder, bool enable = true)
+        {
+            Assert.IsNotNull(pluginBuilder);
```

Comment:
Asserts in production code looks a little strange for me, Should we keep it here?